### PR TITLE
Add support for back link navigation

### PR DESCRIPTION
**Pull request by hadrienk**
This PR adds support for reverse traversal of the links between domain objects. Considering the size of this PR, I suggest we squash it before merging. It is also such a large change that I think we should bump to version 1.0 (maybe go for semver as well?). 

## New persistence API method

The persistence API is updated with a new `readSourceDocuments` method used in a new fetcher:
https://github.com/statisticsnorway/linked-data-store-core/blob/e95efa76ea6800301cb44cd8444ac43304b92912/src/main/java/no/ssb/lds/graphql/fetcher/PersistenceReverseLinksConnectionFetcher.java#L57-L68

## New specification transformation

The transformation of the specification to GraphQL has been redesigned so that we convert to `SDLDefinition`s as a first step. This will later on allow us to rely only on a GraphQL "specification". 
https://github.com/statisticsnorway/linked-data-store-core/blob/e95efa76ea6800301cb44cd8444ac43304b92912/src/main/java/no/ssb/lds/graphql/schemas/SpecificationConverter.java#L69-L74

## Schema processing 

The process of converting the specification/graphql schema now follows a pipeline pattern. The GraphQLSchema is gradually transformed by tree visitors to add types, connection and reverse links. Types and fields are annotated with directives that the next visitors in the pipeline can recognize and act upon. 

https://github.com/statisticsnorway/linked-data-store-core/blob/e95efa76ea6800301cb44cd8444ac43304b92912/src/main/java/no/ssb/lds/graphql/schemas/GraphQLSchemaBuilder.java#L127-L137

1. **Type collection; collect all the types in a map**
```
type Domain @domain {
  linkProperty: OtherDomain @link
}
type OtherDomain @domain {
}
```
2. **Add directives manually until the GSIM team merges the reverseLink annotations**
```
type Domain @domain {
  linkProperty: OtherDomain @link(reverseName: "parentDomains")
}
type OtherDomain @domain {
}
```
3. **Build the reverse links**
```
type Domain @domain {
  linkProperty: OtherDomain @link
}
type OtherDomain @domain {
  parentDomains: [Domain] @reverseLink(mappedBy: "$.linkProperty")
}
```
4. **Build the query fields for the `@domain` types**
```
type Domain @domain {
  linkProperty: OtherDomain @link
}
type OtherDomain @domain {
  parentDomains: [Domain] @reverseLink(mappedBy: "$.linkProperty")
}
type Query {
  DomainById(id: ID!): Domain @link
  Domain: [Domain] @link
  OtherDomainById(id: ID!): OtherDomain @link
  OtherDomain: [OtherDomain] @link
}
```
5. **Add search types and query method**
```
type Domain @domain {
  linkProperty: OtherDomain @link
}
type OtherDomain @domain {
  parentDomains: [Domain] @reverseLink(mappedBy: "$.linkProperty")
}
union SearchResult = Domain | OtherDomain
type Query {
  DomainById(id: ID!): Domain @link
  Domain: [Domain] @link
  OtherDomainById(id: ID!): OtherDomain @link
  OtherDomain: [OtherDomain] @link
  Search(query: String): [SearchType] @link @search
}
```
6. **Replace types with type references**
No change to the schema. This ensure we only have on type definition per object.
7. **Transform links to connections**
``` 
type Domain @domain {
  linkProperty: OtherDomain @link
}
type OtherDomain @domain {
  parentDomains(first: Int, ...): [DomainConnection] @reverseLink(mappedBy: "$.linkProperty")
}
union SearchResult = Domain | OtherDomain
type Query {
  DomainById(id: ID!): Domain @link
  Domain(first: Int, ...): [DomainConnection] @link
  OtherDomainById(id: ID!): OtherDomain @link
  OtherDomain(first: Int, ...): [OtherDomainConnection] @link
  Search(query: String): [SearchTypeConnection] @link @search
}
```
8. **Setup registry with fetchers**
No change to the schema. Create and setup fetchers.

### CODE REVIEW

**Comment by kimcs** [comment]
I would like the **Build the reverse links** section to be done automatically for all links according to a well-defined reverse-link naming convention. I propose the following:
```
 type Domain @domain {
   linkProperty: OtherDomain @link
 }
 type OtherDomain @domain {
   reverseDomainLinkProperty: [Domain] @reverseLink(mappedBy: "$.linkProperty")
 }
```

File: pom.xml [hunk 1]
```diff
@@ -233,14 +245,14 @@
             <groupId>ch.qos.logback</groupId>
             <artifactId>logback-core</artifactId>
             <version>${logback.version}</version>
-            <scope>test</scope>
+            <!--<scope>test</scope>-->
```
Comment (kimcs) [inline pom.xml]: Dette skal ligge i test-scope.

File: pom.xml [hunk 2]
```diff
@@ -233,14 +245,14 @@
...
         </dependency>
 
         <dependency>
             <groupId>ch.qos.logback</groupId>
             <artifactId>logback-classic</artifactId>
             <version>${logback.version}</version>
-            <scope>test</scope>
+            <!--<scope>test</scope>-->
```
Comment (kimcs) [inline pom.xml]: Denne også i test-scope

File: src/test/java/no/ssb/lds/core/domain/reference/ReferenceResourceHandlerTest.java [hunk 3]
```diff
@@ -152,8 +152,8 @@ public void thatPUTLongResourcePathWorks() {
 
     @Test
     public void thatGETLongResourcePathWorks() {
-        client.delete("/data/provisionagreement/r15?sync=true").expectAnyOf(200, 204);
-        client.get("/data/provisionagreement/r15/support/technicalSupport/contact/s2").expect404NotFound();
+        client.delete("/data/provisionagreement/r14?sync=true").expectAnyOf(200, 204);
```
Comment (kimcs) [inline src/test/java/no/ssb/lds/core/domain/reference/ReferenceResourceHandlerTest.java]: Denne må revertes tilbake til r15 slik den var. Meningen her er å ha kjent state før vi opprettet ny ressurs.

Comment (hadrienk) [reply src/test/java/no/ssb/lds/core/domain/reference/ReferenceResourceHandlerTest.java]: Solved.

File: pom.xml (hunk 2 above)
Comment (hadrienk) [reply pom.xml]: Thanks. Solved.

File: pom.xml (hunk 1 above)
Comment (hadrienk) [reply pom.xml]: Thanks. Solved.

**Comment by hadrienk** [comment]
> I would like the **Build the reverse links** section to be done automatically for all links according to a well-defined reverse-link naming convention. I propose the following:
> 
> ```
>  type Domain @domain {
>    linkProperty: OtherDomain @link
>  }
>  type OtherDomain @domain {
>    reverseDomainLinkProperty: [Domain] @reverseLink(mappedBy: "$.linkProperty")
>  }
> ```

Agree. I am gonna add a new PR for this one.

**Pull request merged**